### DISCUSSION TITLE
Proposal: Unnested Local Functions

### DISCUSSION THREAD

**Discussion by ghost** [dotnet/csharplang#1329] [Language Ideas]
I used local functions in practice (as the CopyVideo() function inside the btnSave_Click here:
https://github.com/MohammadHamdyGhanem/VASE/blob/master/src/Forms/VideoEditSaver.cs
I found it the best solution in my situation, but it is still confusing. It is not a good organization to put a full a function among other lines of code. So, I suggest to allow us to re-write this:

```
void foo1()
{
    //...................
    void foo2()
     {

     }
    // ..............
}
```

using "local" keyword as this:
```
void foo1()
{
    //...................
    local void foo2();
    // ..............
}

local void foo1.foo2()
{

}
```

Note:
foo2 still can not be called from any method other than foo1. This is just away to organize the code in the editor, and doesn't affect how IL is generated. All what C# will do is to replace the local function definition line with its imlemetation body! Something like partial methods.
The name  foo1.foo2 should appear in method names dropdown list (in the top of the code editor page).

Properties can be treated in the same way:
```
public int MyProperty {
            get
            {
                //......
                   local void foo3();
                //.....
            }

            set
            {
                //......
                local void foo4();
                //.....           
            }
}

local void MyProperty.foo3()
{
}

local void MyProperty.foo4()
{
}

```

Constructors can be refered to as the name of the class like this:

```
class test
{
    public test()
    {
       // ...............
        int foo5()
         {

        }
       // .................
     }
}
```

This can be re-written as:

```
class test
{
    public test()
    {
       // ...............
        local int foo5();
       // .................
     }
     
     local test.foo5()
     {

     }

}
```

I think it will make the code more organized and easier to read, with the same functionality.

**Comment by alrz** [comment]
you won't be able to capture any variable in that context, therefore I guess  it wouldn't be much different from a regular method.

**Comment by MkazemAkhgary** [comment]
I think it can conflict with explicit interface implements. 

on the other hand, you can write a regular method. you get pretty much same thing.

**Comment by ghost** [comment]
@alrz 
No, The 2 syntaxes do exact the same thing. foo1.foo2 tells the compiler to treat foo2() as a local function inside foo1(), and should generate the same IL code for both.

@MkazemAkhgary 
I don't think there will be any confusion, because foo1 is not an interface name.
But, as a precaution, we can add a "local" keyword as modifier for local methods, like this:

```
local void foo1.foo2()
{

}
```

> you can write a regular method. you get pretty much same thing.

This is not the place to discuss using local functions or not. C# added local functions already, and I suggest an alternative syntax for them. In the code I linked, I put the local function inside #region and but comments and did every thing to make the code looks clear in the editor, but I still feel something is not in place. At least I can't find this function easily, because it doesn't appear in method names dropdown list.

**Comment by svick** [comment]
> foo1.foo2 tells the compiler to treat foo2() as a local function inside foo1()

*Where* inside of `foo1`? It matters, because the local function can access different variables based on its position in the method. Probably the most reasonable choice would be to treat it as a local function that's at the end of the method, so it would be able to access all top-level variables, but I think it would still be confusing.

> I don't think there will be any confusion, because foo1 is not an interface name.

How do you write an unnested "local" function for a method with more than one overload? Other potentially problematic cases: an unnested "local" function for a property accessor, a constructor or an explicit interface implementation.

> I can't find this function easily, because it doesn't appear in method names dropdown list.

If you want the IDE to behave differently, you should propose that the IDE should behave differently, not that the language should change. Especially because changing the IDE is much easier than changing the language.

**Comment by aluanhaddad** [comment]
The whole point of a local function is that it's local. What this is suggesting is just a private method with new and confusing syntax.

If you want to define it in another scope there's already plenty of syntax for that

```c#
class A
{
    void Foo1()
    {
        Foo1Locals.Foo2();
    }

    static class Foo1Locals
    {
        internal static void Foo2() {} 
    }
}
```

**Comment by MkazemAkhgary** [comment]
>If you want the IDE to behave differently, you should propose that the IDE should behave differently.

Totally agree. @MohammadHamdyGhanem Note that these are new language features and it takes some time to for IDE to amend and cope with these new features.

**Comment by bondsbw** [comment]
I could see some benefit in being able to define a local function in a partial class.

```c#
partial class Foo
{
    public void DoSomething()
    {
        // ...
        var items = EnumerateItems();
        // ...

        IEnumerable<Item> EnumerateItems();
    }
}
```

```c#
partial class Foo
{
    local IEnumerable<Item> DoSomething().EnumerateItems()
    {
        // implementation
    }
}
```

Here I explicitly specified the placement of the local method, but perhaps there could be a default location as mentioned in comments above.

I also disambiguated the method signature, in case of overloads.

**Comment by ghost** [comment]
> How do you write an unnested "local" function for a method with more than one overload?

```
local void foo1.foo2()
{

}

local void foo1.foo2(int x)
{

}
```

**Comment by ghost** [comment]
@svick 
I thought of your situations, and you are right. I suggest to use another approach, be adding the method header in its local place, and seperate the implementation be the local unnested idea, like this:

```
public int MyProperty {
            get
            {
                //......
                   local void foo3();
                //.....
            }

            set
            {
                //......
                local void foo4();
                //.....           
            }
}

local void MyProperty.foo3()
{
}

local void MyProperty.foo4()
{
}
```

**Comment by svick** [comment]
@MohammadHamdyGhanem You misunderstood. I'm not asking how to define overloads of a local function (which is not allowed for local functions in the first place). What I'm asking is:

```c#
void foo1() {}

void foo1(int i)
{
    // how do I change this local function to unnested?
    void foo2() {}
}
```

You can probably invent some syntax to reference overloads, but it won't be trivial and the language designers seem to be reluctant to do it (see https://github.com/dotnet/csharplang/issues/712 and [this article](https://blogs.msdn.microsoft.com/ericlippert/2009/05/21/in-foof-we-trust-a-dialogue/)).

**Comment by ghost** [comment]
@svick 
I got it and solved it in the next reply, by adding the local header definition.
The constructors can be refered to as the name of the class like this:

```
class test
{
    public test()
    {
       // ...............
        int foo5()
         {

        }
       // .................
     }
}
```

This can be re-written as:

```
class test
{
    public test()
    {
       // ...............
        local int foo5();
       // .................
     }
     
     local test.foo5()
     {

     }

}
```

**Comment by ghost** [comment]
@bondsbw 
Sorry, I didn't see your replay. I came to close solution by using the keyword local in front of the local method header definition. You can put the method implementation wherever you want.

**Comment by jnm2** [comment]
I don't think anything outside of a method's scope should be able to access the methods locals, arguments, or local functions. Too confusing for zero benefit. If it's complex enough to deserve strict organization, it's too complex to use local functions for and you should move it to its own class.

**Comment by ghost** [comment]
> I don't think anything outside of a method's scope should be able to access the methods locals, arguments, or local functions

nothing is out the scope. Please note this is just away to organize the code in the editor, and doesn't affect how IL is generated. All what C# will do is to replace the local function definition line with its imlemetation body! Something like partial methods.

**Comment by jnm2** [comment]
> nothing is out the scope. 

It is lexically out of scope. The method scope ends with `}`. Once the scope ends, you should not be able to access things declared in that scope. It's too unusual and confusing and truly unnecessary.

**Comment by ghost** [comment]
>  Too confusing for zero benefit

After using local functions in practice, I regreted doing so! I end up with complex tightly coubled un-reuseable code! After ValueTuples, I think I should re-write this function as regular function.
So, the zero benefit comes at first from this local functions, which is similar to the  ABANDONED VB6 GoSub statement. unnesting local function may make it more attractive!

**Comment by MkazemAkhgary** [comment]
>I think it will make the code more organized and easier to read,

Its opinion based. Imho it makes it less organized (methods scattered around are harder to find).

personally I use local function when its much smaller compared to outer method, other wise I prefer one solid method or extension methods.

iterators are exception to this. main method of iterator is around 4 lines to validate inputs. then it calls local function iterator that can be any size.

**Comment by ghost** [comment]
> It is lexically out of scope. The method scope ends with }. Once the scope ends, you should not be able to access things declared in that scope. It's too unusual and confusing and truly unnecessary.

This is still true. The true language of .NET is IL, and c# will deliver a standard IL code. C#, VB and F# are deffernt syntaxes to generate IL. This gives them the freedom to add any high-level features we want.

**Comment by jnm2** [comment]
@MohammadHamdyGhanem You aren't addressing my statement. Of course it's higher-level. Of course it *can* work. I didn't say it can't work; I said it would be a kludge.

My statement was that access to the contents of methods which are lexically out of scope is confusing and that writing a class is superior. Why do you say it is not confusing, and why do you say that organizing it into a class is inferior to this new confusing syntax?

**Comment by bondsbw** [comment]
FWIW partial classes are an example of lexical scope extending beyond the end brace.

**Comment by jnm2** [comment]
@bondsbw They certainly are. They require the `partial` keyword and it does not matter what order members go in. (Except for static initializers, which can be confusing and error-prone.)

Methods are not like classes. As soon as you start trying to treat one like I class, I'm going to suggest that you do your future readers a favor, come up with a good name, and write a class. 😈

**Comment by jnm2** [comment]
@MohammadHamdyGhanem Just tried it. The breakpoint works on that line for me. Are you optimizing?

**Comment by svick** [comment]
> After using local functions in practice, I regreted doing so! I end up with complex tightly coubled un-reuseable code!

Using a feature too much is part of the normal process when a person or community learns about a new feature. It doesn't necessarily mean the feature is bad.

And I don't see how your proposal addresses any of your complaints: It won't help you write code that's less complex or less tightly coupled. And it won't make your code reusable, since unnested local function is still local to a single method. Sure, converting unnested local function to a private method manually might be somewhat easier than converting a normal local function. But in both cases, it's complex enough that an IDE-provided refactoring is a better choice anyway.

**Comment by svick** [comment]
> By the way, the compiler doesn't stop at breakpoints in local functions and doesn't step into them. Is there a solution or did anyone report this?

I don't see how that's relevant to this issue or even this repo. I suggest you create a new issue in the dotnet/roslyn repo about that (ideally including much more details about what you did and how it failed).

**Comment by ghost** [comment]
> And I don't see how your proposal addresses any of your complaints

It doesn't. This is out of the proposal as a response to your comment.

**Comment by CyrusNajmabadi** [comment]
> After using local functions in practice, I regreted doing so! I end up with complex tightly coubled un-reuseable code! 

I don't understand how this proposal addresses that.  You're still using local functions, you're just putting their body somewhere else.

**Comment by Neme12** [comment]
> After using local functions in practice, I regreted doing so! I end up with complex tightly coubled un-reuseable code!

ok... so don't use them?

**Comment by ghost** [comment]
> ok... so don't use them?

Had to then. I want to make my existing code more organized.

**Comment by CyrusNajmabadi** [comment]
> Had to then. I want to make my existing code more organized.

Again, i don't see how moving the local function delcaration to be a sibling addresses your problems 'with complex tightly coubled un-reuseable code!'

The code is just as tightly coupled as before.  The code is just as unreusable as before.  All that has changed is that you've moved the location of the code to a slightly different location.

**Comment by CyrusNajmabadi** [comment]
> Had to then

I don't understand this comment.  I can't imagine a single case where someone would "have to" use local functions.  Indeed, we had the language for 15ish years without local functions, and life was just fine.  You would only use local functions by choice, not because of any compulsion :)

**Comment by aluanhaddad** [comment]
It comes down to a combination of organizational preference and problem domain. As much as I like having local functions in the language, I don't use them as much as I expected to.

C# is a very expressive language and, for what really is a very minimal amount of syntactic pain, it has allowed us to leverage lexical scope  to our hearts content for well over a decade 😀
```C#
IEnumerable<(Action<T>, int)> ThingsWeMightDoInTheFuture<T>(IEnumerable<T> elements) =>
    from outer in elements
    join inner in elements
    on outer.Key equals inner.Key into correlated
    let merge = (Func<T, int>)(x => correlated.Count(e => e.Equals(x)))
    select ((Action<T>)(e => e.Rank = merge(e)), correlated.Aggregate(1, (a, e) => a * e.Key));
```
In other words, local functions have not changed how I code all that much and you certainly don't need to use them to code yourself up a mess of lexically nested scopes.

**Comment by ghost** [comment]
> You would only use local functions by choice, not because of any compulsion :)

If foo2() works on many variables of  foo1() and makes changes to them, then using a local function is faster than declaring a regular function with too many ref params. Did you use a function with 20 params before?
Local functions are not entirely new, as they could be simulated by using 2 goto commands! Local functions bring us back to the dark ages of goto (and GoSub in VB6)! 
In fact I felt lazy to declare ref params and was exited to try a local function. I will never do that agian :)

**Comment by CyrusNajmabadi** [comment]
Alright, then i don't know what your position is :)    You made some claims, but haven't explained them.  So i'm really unsure what the problem is, or why this proposal is a suitable approach to address whatever that problem actually is :-/

**Comment by ghost** [comment]
The proposal is what is written in the first post. No more.

> I used local functions in practice (as the CopyVideo() function inside the btnSave_Click here:
> https://github.com/MohammadHamdyGhanem/VASE/blob/master/src/Forms/VideoEditSaver.cs
> I found it the best solution in my situation, but it is still confusing. It is not a good organization to put a full a function among other lines of code. So, I suggest to allow us to re-write this:
> 
> ```
> void foo1()
> {
>     //...................
>     void foo2()
>      {
> 
>      }
>     // ..............
> }
> ```
> 
> using "local" keyword as this:
> ```
> void foo1()
> {
>     //...................
>     local void foo2();
>     // ..............
> }
> 
> local void foo1.foo2()
> {
> 
> }
> ```
> 
> Note:
> foo2 still can not be called from any method other than foo1. This is just away to organize the code in the editor, and doesn't affect how IL is generated. All what C# will do is to replace the local function definition line with its imlemetation body! Something like partial methods.
> The name  foo1.foo2 should appear in method names dropdown list (in the top of the code editor page).
> 
> Properties can be treated in the same way:
> ```
> public int MyProperty {
>             get
>             {
>                 //......
>                    local void foo3();
>                 //.....
>             }
> 
>             set
>             {
>                 //......
>                 local void foo4();
>                 //.....           
>             }
> }
> 
> local void MyProperty.foo3()
> {
> }
> 
> local void MyProperty.foo4()
> {
> }
> 
> ```
> 
> Constructors can be refered to as the name of the class like this:
> 
> ```
> class test
> {
>     public test()
>     {
>        // ...............
>         int foo5()
>          {
> 
>         }
>        // .................
>      }
> }
> ```
> 
> This can be re-written as:
> 
> ```
> class test
> {
>     public test()
>     {
>        // ...............
>         local int foo5();
>        // .................
>      }
>      
>      local test.foo5()
>      {
> 
>      }
> 
> }
> ```
> 
> I think it will make the code more organized and easier to read, with the same functionality.

**Comment by CyrusNajmabadi** [comment]
> Something like partial methods.

This is not a good justification/example.  The purpose of partial methods is to allow scenarios where one producer of an API defines an extension point that someone else can fill in.  They can be not filled in, and then have no impact on what's compiled.

There is no such reason/benefit to your proposal.  I'm not sure what the benefit of this is or what problem it is addressing.  It just seems to make things more confusing.  it moves the code away from where you need it.

If you want to write things this way then just write things like this:

```c#
    public test()
    {
       var foo5 = test_foo5();
     }
     
     Func<int> test_foo5()
     {
         return () => ...
     }
```

Finally, these examples are completely non-compelling.  Your problem example doens't seem like a problme, and your solution doesn't demonstrate why this would be desirable.  Please provide realistic examples to help drive home the case.

For example, see this PR as one that shows a much more realistic example that demonstrates why a solution is needed: https://github.com/dotnet/csharplang/issues/1328

**Comment by DavidArno** [comment]
> If foo2() works on many variables of foo1() and makes changes to them, then using a local function is faster than declaring a regular function with too many ref params. Did you use a function with 20 params before?
Local functions are not entirely new, as they could be simulated by using 2 goto commands! Local functions bring us back to the dark ages of goto (and GoSub in VB6)!
In fact I felt lazy to declare ref params and was exited to try a local function. I will never do that agian :)

You have a function, `foo1` that has at least 20 mutable variables. There's your problem, right there. `foo1` is trying to do too much. Unfortunately, you appear to have tried to fix it by pushing some of the functionality off to `foo2`. As a result, you've have to use `ref` parameters to pass all those variables into that section of `foo1` that you moved to `foo2`. Along comes local functions and you think your problems solved, as you then put `foo2` back into `foo1`, removing the need for all those parameters. But the code is now back in `foo1` and you back to the huge sprawling mess you had at the start, except that some of the code is in a "sub function".

So now, you are thinking, "well if I could move those local functions outside of `foo1`, I get the best of both worlds: `foo1` becomes smaller and I don't need those 20 ref params. But what you now have is direct use of the [action at a distance anti-pattern](https://en.wikipedia.org/wiki/Action_at_a_distance_(computer_programming)).

All along, you are ignoring the elephant in the room: those 20 variables. They have been telling you from the start that you are trying to do too much in one function. You need to reduce them. There is a range of solutions you could employ. At one end of that spectrum lies the "ideal" solution: rewrite the entire function to simplify it. Of course to confidently do this, you'll need a solid set of unit tests backing you up. Apologies if I'm misjudging you, but I'm going to guess that you don't have unit tests. 

At the other end of the spectrum is "fudge it". Take those 20 variables, and make them fields in your class. Normally, I'd rather chew off my own right hand than suggest someone do this, but I think it likely the best solution in your case. You'll still be using "action at a distance", but it'll let you chop `foo` up into `foo1` to `foo99` a and will remove the need for those 20 ref parameters. Plus it is an already available feature of the language.  

Local functions are a useful addition to the language. Please do not confuse your attempted misuse of them with signs that they are a bad idea. Thanks.

**Comment by ghost** [comment]
@DavidArno
If you look at Matrix4x4.Invert method here
https://github.com/dotnet/corefx/blob/master/src/System.Numerics.Vectors/src/System/Numerics/Matrix4x4.cs
You'll see at least 27 floot variables defind! Each job decides how much variables it needs.
The 20 ref variables is just an example, but it could happen and you couldn't divide the function into smaller ones. 
You are free to use local functions as you need, I didn't ask to remove them, but I want to be able to move their code into a separate block if it is large, or if the parent function is complicated. This means I can write them as nested or untested local functions. Each job decides to use lambdas, delegates or local functions. But as I said, using local functions brought me 20 years back to remind me of some mistakes I have forgotten!

**Comment by DavidArno** [comment]
> If you look at Matrix4x4.Invert method here ... You'll see at least 27 floot variables defind

If I look at, eg `public static bool Invert(Matrix4x4 matrix, out Matrix4x4 result)`, I sure do see lots of variables. And yet the code is clear and easy to follow, despite being written with performance, rather than elegance, in mind. They are almost all used as write-once temporary value stores and the code isn't complicated. The author has done an excellent job of keeping the code easy to read therefore. This proves that many variables can be used just fine.

But if you find yourself thrashing around trying to shoehorn your code into inappropriate features because it's long, complicated and unreadable, then those many variables clearly aren't fine. There's no hard and fast rule on this, but if you have a local function that's more then say half dozen lines long, then alarm bells should ring that maybe a local function isn't the right solution.

**Comment by AustinBryan** [comment]
This is actually exactly what the compiler already does. 

```csharp
public static void Main() 
{
     int Foo() => 3;
}

// Becomes:
public static class Program
{
    public static void Main()
    {
    }

    [CompilerGenerated]
    internal static int <Main>g__Foo|0_0()
    {
        return 3;
    }
}
```
Unlike nested functions in JavaScript, Local Functions aren't actually contained in the outer function, they get moved outside. This makes them more efficient.

So, what I don't get, is what are we gaining in doing it in your syntax? Local functions are already organized, *inside* the function they belong in. We shouldn't have to see the function outside of it because it will make things more complicated.

If you're worried about your outer method being really huge now, that's fine. Jon Skeet said his methods with local functions get big and it looks like it needs to be refactored, except it already is, and everything is nicely scoped. 

You could do:

```csharp
public static void Main() 
{
#region Local Functions
     int Foo() => 3;
#endregion Local Functions
}

```

**Comment by DavidArno** [comment]
@AustinBryan,

To my mind, putting a `#region` *inside* a method constitutes gross misconduct and is grounds for instant dismissal... 😡
